### PR TITLE
feat: add support of multiple messages into transaction handler

### PR DESCRIPTION
**Pull request by dmvict**

### CODE REVIEW

File: gears/src/types/tx/mod.rs [hunk 1]
```diff
@@ -26,6 +26,35 @@ pub trait TxMessage:
     fn type_url(&self) -> &'static str;
 }
 
+/// Utility type that guarantees correctness of transaction messages set
+pub struct Messages<T: TxMessage>(pub(crate) Vec<T>);
```
Comment (joneskm) [inline gears/src/types/tx/mod.rs:30]: Do we really want this public across the entire crate?

File: gears/src/types/tx/mod.rs [hunk 2]
```diff
@@ -26,6 +26,35 @@ pub trait TxMessage:
...
+
+impl<T: TxMessage> Messages<T> {
+    pub fn new(messages: Vec<T>) -> anyhow::Result<Messages<T>> {
+        if messages.is_empty() {
+            Err(anyhow::anyhow!(
+                "transaction applies non empty set of messages"
```
Comment (joneskm) [inline gears/src/types/tx/mod.rs:36]: How about using a custom error for this?

File: gears/src/types/tx/mod.rs (hunk 1 above)
Comment (dmvict) [reply gears/src/types/tx/mod.rs]: Used a function to convert the instance.

File: gears/src/types/tx/mod.rs (hunk 2 above)
Comment (dmvict) [reply gears/src/types/tx/mod.rs]: Added custom error.

File: gears/src/types/tx/mod.rs [hunk 3]
```diff
@@ -26,6 +29,53 @@ pub trait TxMessage:
     fn type_url(&self) -> &'static str;
 }
 
+/// Utility type that guarantees correctness of transaction messages set
+pub struct Messages<T: TxMessage> {
+    messages: Vec<T>,
+    /// A number of messages in the transaction. Zero means unlimited number of messages.
+    /// Default is 0
+    chunk_size: u16,
```
Comment (joneskm) [inline gears/src/types/tx/mod.rs:37]: Does it make sense to make this a usize from the get go?

File: gears/src/commands/client/tx.rs [hunk 4]
```diff
@@ -39,31 +85,78 @@ pub struct LocalInfo {
...
                 info.keyring_backend.to_keyring_backend(&keyring_home),
             )?;
 
-            let message = handler.prepare_tx(inner, key.get_address())?;
-            handler.handle_tx(message, key, node, chain_id, fees, SignMode::Direct)
+            let ctx = &(&command).into();
+            let messages = handler.prepare_tx(ctx, command.inner, key.get_address())?;
+
+            if messages.chunk_size() > 0
+            /* && command.broadcast_mode == BroadcastMode::Block */
```
Comment (joneskm) [inline gears/src/commands/client/tx.rs:125]: Did you forget to remove?

File: gears/src/commands/client/tx.rs [hunk 5]
```diff
@@ -39,31 +85,78 @@ pub struct LocalInfo {
...
+            {
+                let step = messages.chunk_size() as usize;
+                let msgs = messages.into_msgs();
+
+                let mut res = vec![];
+                for i in (0..msgs.len()).step_by(step) {
+                    res.push(
+                        handler.handle_tx(
+                            msgs[i..(i + step).min(msgs.len())]
```
Comment (joneskm) [inline gears/src/commands/client/tx.rs:134]: Would [this](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.chunks) be useful here?

Comment (dmvict) [reply gears/src/commands/client/tx.rs:134]: It's exactly what I've searched in editor popup tips and didn't find.

File: gears/src/commands/client/tx.rs (hunk 4 above)
Comment (dmvict) [reply gears/src/commands/client/tx.rs:125]: I forget to add todo. The feature with broadcast mode is not implemented yet.

File: gears/src/types/tx/mod.rs (hunk 3 above)
Comment (dmvict) [reply gears/src/types/tx/mod.rs:37]: I think yes. Thank you.

File: gears/src/types/tx/errors.rs [hunk 6]
```diff
@@ -1,5 +1,11 @@
 use crate::types::auth::info::AuthError;
 
+#[derive(Debug, thiserror::Error)]
+pub enum MessagesError {
+    #[error("transaction applies non empty set of messages")]
```
[the remark this answers is not in the document]
Comment (dmvict) [reply gears/src/types/tx/errors.rs:5]: Updated

File: gears/src/commands/client/tx.rs [hunk 7]
```diff
@@ -39,31 +85,78 @@ pub struct LocalInfo {
...
             let key = LedgerProxyKey::new()?;
 
-            let message = handler.prepare_tx(inner, key.get_address())?;
-            handler.handle_tx(message, key, node, chain_id, fees, SignMode::Textual)
+            let ctx = &(&command).into();
+            let messages = handler.prepare_tx(ctx, command.inner, key.get_address())?;
+            // TODO: maybe remove error and replace with warning message
+            if messages.chunk_size() != 0 {
+                return Err(anyhow!(
+                    "ledger device cannot create more that one transaction per command run"
```
[the remark this answers is not in the document]
Comment (dmvict) [reply gears/src/commands/client/tx.rs]: Added borrowing of the key.

**Review by NYBACHOK — APPROVED** [review]

**Pull request merged**